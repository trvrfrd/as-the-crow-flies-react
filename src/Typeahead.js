import React, { Component } from 'react';
import { escapeRegExp } from './utils';

export default class Typeahead extends Component {
  state = {
    query: '',
    results: [],
    showSuggestions: false
  }

  handleChange = e => this.handleQuery(e.target.value);

  handleQuery = query => {
    // by definition, there can't be a selection immediately after query input
    // so: make sure parent knows selection is null
    this.props.onSelect(null);

    if (query.length === 0) return this.handleBlankQuery();

    const results = this.getQueryResults(query);
    this.setState({
      query,
      results,
      showSuggestions: true
    });
  }

  handleBlankQuery = () => {
    this.setState({
      query: '',
      results: [],
      showSuggestions: false
    })
  }

  handleSelect = data => {
    this.props.onSelect(data);
    this.setState({
      showSuggestions: false,
      query: this.props.formatSuggestion(data)
    });
  }

  getQueryResults = query => {
    const regexp = new RegExp(escapeRegExp(query), 'i');
    return this.props.source.filter(data =>
      Object.values(data).some(attr => regexp.test(attr))
    ).slice(0, this.props.maxSuggestions);
  }

  highlightQuery = str => {
    const regexp = new RegExp(escapeRegExp(this.state.query), 'ig');
    const __html = str.replace(regexp, match => `<span class="highlight">${match}</span>`);
    return { __html };
  }

  render() {
    const { name, placeholder } = this.props;
    const { results } = this.state;

    return (
      <div className="typeahead">

        <fieldset>
          <label htmlFor={name}>{name}: </label>
          <br />
          <input
            id={name}
            name={name}
            placeholder={placeholder}
            value={this.state.query}
            onChange={this.handleChange}
          />
        </fieldset>

        {
          this.state.showSuggestions
            ?
            <div className="suggestions">
              {results.map((data, idx) =>
                <div
                  className="suggestion"
                  key={idx}
                  onClick={() => this.handleSelect(data)}
                  dangerouslySetInnerHTML={this.highlightQuery(this.props.formatSuggestion(data))}
                />
              )}
            </div>
            : null
        }

      </div>
    )
  }
}

Typeahead.defaultProps = {
  maxSuggestions: 10
};
