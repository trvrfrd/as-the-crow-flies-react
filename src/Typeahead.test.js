import React from 'react';
import { shallow } from 'enzyme';
import Typeahead from './Typeahead';

const formatFunction = JSON.stringify;

test('Typeahead basic rendering snapshot', () => {
  const wrapper = shallow(<Typeahead />);
  expect(wrapper).toMatchSnapshot();
});
