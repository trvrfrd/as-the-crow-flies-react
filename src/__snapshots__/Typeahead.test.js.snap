// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typeahead basic rendering snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Typeahead />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <fieldset>
          <label>
            : 
          </label>
          <br />
          <input
            onChange={[Function]}
            value=""
          />
        </fieldset>,
        null,
      ],
      "className": "typeahead",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <label>
              : 
            </label>,
            <br />,
            <input
              onChange={[Function]}
              value=""
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                undefined,
                ": ",
              ],
              "htmlFor": undefined,
            },
            "ref": null,
            "rendered": Array [
              undefined,
              ": ",
            ],
            "type": "label",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "br",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "id": undefined,
              "name": undefined,
              "onChange": [Function],
              "placeholder": undefined,
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
        ],
        "type": "fieldset",
      },
      null,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <fieldset>
            <label>
              : 
            </label>
            <br />
            <input
              onChange={[Function]}
              value=""
            />
          </fieldset>,
          null,
        ],
        "className": "typeahead",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <label>
                : 
              </label>,
              <br />,
              <input
                onChange={[Function]}
                value=""
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  undefined,
                  ": ",
                ],
                "htmlFor": undefined,
              },
              "ref": null,
              "rendered": Array [
                undefined,
                ": ",
              ],
              "type": "label",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": "br",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "id": undefined,
                "name": undefined,
                "onChange": [Function],
                "placeholder": undefined,
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
          ],
          "type": "fieldset",
        },
        null,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
